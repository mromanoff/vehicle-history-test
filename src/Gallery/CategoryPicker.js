import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '../components/Button';

class CategoryPicker extends Component {
  static propTypes = {
    categories: PropTypes.array.isRequired,
    currentCategory: PropTypes.string.isRequired,
    onCategoryChange: PropTypes.func.isRequired,
  };

  state = {};

  onClick(category) {
    this.props.onCategoryChange(category);
    this.setState({ category: category });
  }

  componentDidMount() {
    const { onCategoryChange, currentCategory } = this.props;
    if (currentCategory) {
      this.setState({ category: currentCategory });
      onCategoryChange(currentCategory);
    }
  }

  render() {
    return (
      <section>
        <h1 className="title">Pick category</h1>
        <div className="module category-picker">
          {this.props.categories.map(category => {
            return (
              <Button
                key={category}
                className={`btn btn--tertiary btn--sm ${
                  this.state.category === category ? 'btn--tertiary-active' : ''
                }`}
                onClick={() => this.onClick(category)}
              >
                {category}
              </Button>
            );
          })}
        </div>
      </section>
    );
  }
}

export default CategoryPicker;
