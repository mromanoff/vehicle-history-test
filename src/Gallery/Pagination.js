import React from 'react';
import PropTypes from 'prop-types';
import Pagination from 'react-js-pagination';
import styled from 'styled-components';

const StyledPagination = styled.div`
  margin-bottom: 2rem;
  > ul {
    display: flex;
    justify-content: center;
    margin: 0;
    padding: 0;

    > li {
      margin-right: 0.5rem;
      border-radius: 50%;
      background-color: ${props => props.theme.color.grayLight};
      font-size: ${props => props.theme.size.sm};
      color: ${props => props.theme.color.white};
      transition: background-color 100ms ease-out 100ms;
      display: flex;
      align-items: center;
      justify-content: center;
      text-align: center;
      width: 46px;
      height: 46px;
      font-weight: 300;

      &:last-child {
        margin-right: 0;
      }

      &.active {
        color: ${props => props.theme.color.white};
        background-color: ${props => props.theme.color.primary5};
      }

      &:hover {
        background-color: ${props => props.theme.color.primary5};
      }

      &.disabled {
        display: none;
      }

      > a,
      span {
        color: ${props => props.theme.color.white};
        text-decoration: none;
      }
    }
  }
`;

const GalleryPagination = ({
  currentPage,
  itemsCountPerPage,
  totalItemsCount,
  onChange,
}) => {
  const pageCount = Math.ceil(totalItemsCount / itemsCountPerPage);

  return (
    pageCount > 1 && (
      <StyledPagination>
        <Pagination
          activePage={currentPage}
          itemsCountPerPage={itemsCountPerPage}
          totalItemsCount={totalItemsCount}
          pageRangeDisplayed={3}
          onChange={onChange}
        />
      </StyledPagination>
    )
  );
};

export default GalleryPagination;

GalleryPagination.propTypes = {
  currentPage: PropTypes.number.isRequired,
  itemsCountPerPage: PropTypes.number.isRequired,
  totalItemsCount: PropTypes.number.isRequired,
  onChange: PropTypes.func.isRequired,
};
