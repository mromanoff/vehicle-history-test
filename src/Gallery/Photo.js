import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Photo extends Component {
  static propTypes = {
    photo: PropTypes.object.isRequired,
  };

  render() {
    const { images, name } = this.props.photo;
    return (
      <div className="gallery__item">
        <img src={images[0].url} alt={name} className="full-width" />
      </div>
    );
  }
}

export default Photo;
