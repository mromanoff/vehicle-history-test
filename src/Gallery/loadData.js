const URL =
  'https://api.500px.com/v1/photos?feature=popular&image_size=440&exclude=Nude&rpp=20&consumer_key=vRemLRvbgOrkPsJhzeoGdSNHiuC22aZ4TgwgXQXK';

//const request = {
const loadData = async ({category, page}) => {
  const encodedCategory = encodeURIComponent(category);
  const encodedPage = encodeURIComponent(page);

  try {
    const url = `${URL}&page=${encodedPage}&only=${encodedCategory}`;
    const response = await fetch(url);
    return await response.json();
  } catch (error) {
    throw new Error(error);
  }
};

export default loadData;
