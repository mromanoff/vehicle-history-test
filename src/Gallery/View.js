import React from 'react';
import Loader from '../components/Loader';
import Error from '../components/Error';
import Pagination from './Pagination';
import CategoryPicker from './CategoryPicker';
import Photo from './Photo';

const View = ({
  isLoading,
  isError,
  photos,
  error,
  categories,
  currentCategory,
  onCategoryChange,
  onPageChange,
  currentPage,
  totalItemsCount,
  itemsCountPerPage,
  RenderPagination,
  RenderPhoto,
  RenderCategoryPicker,
  RenderLoader,
  RenderError,
}) => (
  <div>
    <RenderCategoryPicker
      categories={categories}
      currentCategory={currentCategory}
      onCategoryChange={onCategoryChange}
    />
    <section>
      {isLoading && <RenderLoader />}
      {isError && <RenderError error={error} />}

      <div className={isLoading ? 'loading' : 'loaded'}>
        {photos && (
          <div>

            <RenderPagination
              currentPage={currentPage}
              totalItemsCount={totalItemsCount}
              itemsCountPerPage={itemsCountPerPage}
              onChange={onPageChange}
            />

            <div className="module gallery">
              {photos.map(photo => (
                <RenderPhoto key={photo.id} photo={photo} />
              ))}
            </div>

            <RenderPagination
              currentPage={currentPage}
              totalItemsCount={totalItemsCount}
              itemsCountPerPage={itemsCountPerPage}
              onChange={onPageChange}
            />

          </div>
        )}
      </div>
    </section>
  </div>
);
View.propTypes = {};

View.defaultProps = {
  RenderCategoryPicker: CategoryPicker,
  RenderPhoto: Photo,
  RenderPagination: Pagination,
  RenderLoader: Loader,
  RenderError: Error,
};

export default View;
