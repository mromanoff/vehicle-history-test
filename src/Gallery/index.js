import React, { Component } from 'react';

import View from './View';
import loadData from './loadData';

class Categories extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: false,
      isError: false,
      error: null,
      currentPage: this.props.initialPage || 1,
      currentCategory: this.props.initialCategory, // || 'All'?
      itemsCountPerPage: this.props.itemsCountPerPage || 20,
    };
  }

  load = async () => {
    try {
      this.setState({ isLoading: true, error: false });
      const data = await loadData({
        category: this.state.currentCategory,
        page: this.state.currentPage
      });
      this.setState({ isLoading: false });
      this.normalizeApi(data);
    } catch (error) {
      this.setState({ isLoading: false, error: true });
    }
  };

  handleCategory = currentCategory => {
    //on category change. set current page to 1.
    this.setState(
      {
        currentCategory,
        currentPage: 1,
      },
      this.load,
    );
  };

  handlePagination = currentPage => {
    this.setState({ currentPage }, this.load);
  };

  normalizeApi = data => {
    this.setState({
      photos: data.photos,
      currentPage: data.current_page,
      totalItemsCount: data.total_items,
    });
  };

  render() {
    return (
      <View
        {...this.props}
        {...this.state}
        onCategoryChange={this.handleCategory}
        onPageChange={this.handlePagination}
      />
    );
  }
}

export default Categories;
