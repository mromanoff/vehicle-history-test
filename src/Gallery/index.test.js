import React from 'react';
import Gallery from './index';

it('Categories renders correctly', () => {
  const wrapper = shallow(<Gallery />);

  //const fetchPhotos = jest.fn();

  expect(wrapper).toMatchSnapshot();
});

