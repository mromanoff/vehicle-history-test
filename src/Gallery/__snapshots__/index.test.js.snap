// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories renders correctly 1`] = `
<View
  RenderCategoryPicker={[Function]}
  RenderError={[Function]}
  RenderLoader={[Function]}
  RenderPagination={[Function]}
  RenderPhoto={[Function]}
  currentPage={1}
  error={null}
  isError={false}
  isLoading={false}
  itemsCountPerPage={20}
  onCategoryChange={[Function]}
  onPageChange={[Function]}
/>
`;
